### PR TITLE
[LETS-136] Disconnect transaction server from page server

### PR DESCRIPTION
**Pull request by adrianbrab**
http://jira.cubrid.org/browse/LETS-136

The transaction server disconnects from the page server when it shuts down. All resources involved in the connection, on the page server, are destroyed.

The transaction server reconnects to the page server and the connection is re-created.

### CODE REVIEW

File: src/base/server_type.cpp [hunk 1]
```diff
@@ -86,11 +86,6 @@ void finalize_server_type ()
     {
       ats_Gl.disconnect_page_server ();
     }
-  else
-    {
-      assert (get_server_type () == SERVER_TYPE_PAGE);
-      ps_Gl.disconnect_active_tran_server ();
-    }
```
Comment (ribram) [inline src/base/server_type.cpp]: This might still have to be done. If the page server is shutting down, it will have to remove the connection with the transaction server.

What is the reasoning for removing the code?

File: src/server/active_tran_server.cpp [hunk 2]
```diff
@@ -276,8 +276,17 @@ void
 active_tran_server::disconnect_page_server ()
 {
   assert_is_active_tran_server ();
+  std::string msg = "" + static_cast<int> (cubcomm::server_server::CONNECT_ACTIVE_TRAN_TO_PAGE_SERVER);
```
Comment (ribram) [inline src/server/active_tran_server.cpp]: Send the memory reprezentation of cubcomm::server_server::CONNECT_ACTIVE_TRAN_TO_PAGE_SERVER.
```suggestion
  const int payload = cubcomm::server_server::CONNECT_ACTIVE_TRAN_TO_PAGE_SERVER;
  std::string msg (reinterpret_cast<const char *> (&payload), sizeof (payload));
```

File: src/server/active_tran_server.cpp [hunk 3]
```diff
@@ -276,8 +276,17 @@ void
...
+  er_log_debug (ARG_FILE_LINE, "Transaction server starts disconnecting from the page servers.");
 
+  for (int i = 0; i < m_page_server_conn_vec.size(); i++)
```
Comment (ribram) [inline src/server/active_tran_server.cpp]:
```suggestion
  for (size_t i = 0; i < m_page_server_conn_vec.size(); i++)
```

File: src/server/active_tran_server.cpp [hunk 4]
```diff
@@ -276,8 +276,17 @@ void
...
+    {
+      er_log_debug (ARG_FILE_LINE, "Transaction server disconnected from page server with channel id: %s.\n",
+		    m_page_server_conn_vec[i]->get_underlying_channel_id ());
+      push_request (i, ats_to_ps_request::SEND_DISCONNECT_MSG, std::move (msg));
```
Comment (ribram) [inline src/server/active_tran_server.cpp]: - Do not change push_request() function. Access m_page_server_conn_vec directly.
- If there are multiple page servers, moving the message on the first request can cause invalid memory access in the other requests. Copy it instead.
```suggestion
      m_page_server_conn_vec[i]->push (ats_to_ps_request::SEND_DISCONNECT_MSG, msg);
```

File: src/server/active_tran_server.hpp [hunk 5]
```diff
@@ -59,7 +59,7 @@ class active_tran_server
 
     bool uses_remote_storage () const;
 
-    void push_request (ats_to_ps_request reqid, std::string &&payload);
+    void push_request (int page_server_index, ats_to_ps_request reqid, std::string &&payload);
```
Comment (ribram) [inline src/server/active_tran_server.hpp]:
```suggestion
    void push_request (ats_to_ps_request reqid, std::string &&payload);
```

File: src/server/ats_ps_request.hpp [hunk 6]
```diff
@@ -23,14 +23,16 @@ enum class ats_to_ps_request
...
+  SEND_DISCONNECT_MSG
 };
 
 enum class ps_to_ats_request
 {
   SEND_SAVED_LSA,
   SEND_LOG_PAGE,
-  SEND_DATA_PAGE
+  SEND_DATA_PAGE,
+  SEND_DISCONNECT_RSP
```
Comment (ribram) [inline src/server/ats_ps_request.hpp]: Not used
```suggestion
  SEND_DATA_PAGE
```

File: src/server/page_server.cpp [hunk 7]
```diff
@@ -124,6 +131,12 @@ void page_server::receive_data_page_fetch (cubpacking::unpacker &upk)
 				   std::placeholders::_2));
 }
 
+void page_server::receive_disconnect_request (cubpacking::unpacker &upk)
+{
+  //start a thread to destroy the ATS to PS connection object
+  std::thread thread (&page_server::disconnect_active_tran_server, std::ref (*this));
```
Comment (cristiarg) [inline src/server/page_server.cpp]: Use another name for the variable - `disconnect_thread`. Thread should be detached or joined.

File: src/base/server_type.cpp (hunk 1 above)
Comment (adrianbrab) [reply src/base/server_type.cpp]: If one PS shuts down shouldn't the ATS be notified? The connection, from the ATS point of view, is still up

File: src/server/active_tran_server.cpp (hunk 3 above)
Comment (adrianbrab) [reply src/server/active_tran_server.cpp]: Changed

**Review by cristiarg — APPROVED** [review]

File: src/server/active_tran_server.cpp (hunk 4 above)
Comment (adrianbrab) [reply src/server/active_tran_server.cpp]: Done.

File: src/server/ats_ps_request.hpp (hunk 6 above)
Comment (adrianbrab) [reply src/server/ats_ps_request.hpp]: Removed now

File: src/server/page_server.cpp (hunk 7 above)
Comment (adrianbrab) [reply src/server/page_server.cpp]: Used detach. Is it really necessary in this context?

Comment (ribram) [reply src/server/page_server.cpp]: https://en.cppreference.com/w/cpp/thread/thread/~thread

This proves @cristiarg is right and it must be detached.

Comment (cristiarg) [reply src/server/page_server.cpp]: Either `join` or `detach` is needed in any context. If merely to express intent. If process decides to close (not crash) in that instant, and the thread did not do its job, process remains hanging (even for a tiny amount of time).

File: src/base/server_type.cpp (hunk 1 above)
Comment (ribram) [reply src/base/server_type.cpp]: It should, and `disconnect_active_tran_server ()` will have to incorporate that functionality. It is not however in the scope of this pull request. and it will have to be updated later. In the meantime, my opinions is that it is better to have it called.

File: src/server/page_server.cpp (hunk 7 above)
Comment (cristiarg) [reply src/server/page_server.cpp]: Actually I was wrong. Without `detach` the thread might not get to do what was intended, and its dtor will cause a core.

File: src/server/page_server.cpp [hunk 8]
```diff
@@ -124,6 +131,13 @@ void page_server::receive_data_page_fetch (cubpacking::unpacker &upk)
 				   std::placeholders::_2));
 }
 
+void page_server::receive_disconnect_request (cubpacking::unpacker &upk)
+{
+  //start a thread to destroy the ATS to PS connection object
+  std::thread disconnect_thread (&page_server::disconnect_active_tran_server, std::ref (*this));
+  disconnect_thread.detach();
```
Comment (ribram) [inline src/server/page_server.cpp]:
```suggestion
  disconnect_thread.detach ();
```

File: src/server/active_tran_server.cpp [hunk 9]
```diff
@@ -276,8 +276,18 @@ void
...
   assert_is_active_tran_server ();
+  const int payload = static_cast<int> (cubcomm::server_server::CONNECT_ACTIVE_TRAN_TO_PAGE_SERVER);
+  std::string msg (reinterpret_cast<const char *> (&payload), sizeof (payload));
+  er_log_debug (ARG_FILE_LINE, "Transaction server starts disconnecting from the page servers.");
 
+  for (size_t i = 0; i < m_page_server_conn_vec.size(); i++)
+    {
+      er_log_debug (ARG_FILE_LINE, "Transaction server disconnected from page server with channel id: %s.\n",
+		    m_page_server_conn_vec[i]->get_underlying_channel_id ());
+      m_page_server_conn_vec[i]->push (ats_to_ps_request::SEND_DISCONNECT_MSG, std::move (msg));
```
Comment (ribram) [inline src/server/active_tran_server.cpp]:
```suggestion
      m_page_server_conn_vec[i]->push (ats_to_ps_request::SEND_DISCONNECT_MSG, std::move (std::string (msg)));
```

**Review by ribram — APPROVED** [review]
Please make the request changes and merge.

File: src/server/page_server.cpp (hunk 8 above)
Comment (adrianbrab) [reply src/server/page_server.cpp]: Fixed now.

File: src/server/active_tran_server.cpp (hunk 9 above)
Comment (adrianbrab) [reply src/server/active_tran_server.cpp]: Done.
